### we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The aim of this project is to clean the data and analyze the included used car listings.

##### The data dictionary provided with data is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.

* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first                              registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first                             registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet                         repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

In [141]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [142]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [143]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

__The above dataset has 20 columns and 50,000 rows.__

__All columns except _1) `vehicleType` 2) `gearbox` 3) `model` 4) `fuelType` 5) `notRepairedDamage` have non-null values for all the rows.__

__There are five columns with `'dtype'` as `'integer'` and rest of the fifteen columns are of `'object'` `'dtype'`.__

__The size of the total data is abobe 7.6 MB.__

_The data taken from the website of used cars is from the year 2016._ 

_There are no pictures provided for the ads of the used cars._



__Let's convert the column names from camelcase to snakecase and re-word some of the column names based on the data dictionary to be more descriptive.__

In [144]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [145]:
# Folloowing map defines the new names for each of the columns.
col_names_map={ 'dateCrawled': 'date_crawled', 
                'name' : 'name', 
                'seller' : 'seller',
                'offerType' : 'offer_type', 
                'price': 'price', 
                'abtest' : 'abtest',
                'vehicleType' : 'vehicle_type',
                'yearOfRegistration' : 'registration_year',
                'gearbox' : 'gearbox', 
                'powerPS' : 'power_PS',
                'model' : 'model',
                'odometer' : 'kilometers_run' ,
                'monthOfRegistration' : 'registration_month',
                'fuelType' : 'fuel_type',
                'brand' : 'brand',
                'notRepairedDamage' : 'unrepaired_damage',
                'dateCreated' : 'ad_created',
                'nrOfPictures' : 'no_of_pics',
                'postalCode' : 'postal_code',
                'lastSeen' : 'last_seen_online'
              }
autos.rename(columns = col_names_map, inplace = True )
print(autos.columns)


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'kilometers_run', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pics', 'postal_code',
       'last_seen_online'],
      dtype='object')


In [146]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model kilometers_run  registration_month fuel_type  \
0    manuell       158  andere      150,000km                   3       lpg   
1  automatik       286     7er      150,000km                   6    benzin   
2    manuell       102    golf       70,000km                   7    benzin   
3  automatik        71  fortwo       70,000km                   6    benzin   
4    manuell         0   focus      150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pics  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

      last_seen_online  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

__We made the following edits to columns names:__

1) `yearOfRegistration` to `registration_year` 

2) `monthOfRegistration` to `registration_month` 

3) `notRepairedDamage` to `unrepaired_damage`

4) `dateCreated` to `ad_created`

5) `odometer` to `kilometers_run`

6) `nrOfPictures` to `no_of_pics`

7) `lastSeen` to `last_seen_online`


_and, the rest of the column names from camelcase to snakecase._

We did the changes for easy readability and which makes more sense to us. Also, the changes are according the Pyhon convention of names. In Python, it is the standard to use snakecase rather than camelcase. 

In [147]:
autos.describe(include='all')
#autos["seller"].unique()
#autos["offer_type"].unique()

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       kilometers_run  registration_month fuel_type       brand  \
count           50000        50000.000000     45518       50000   
unique             13                 NaN         7          40   
top         150,000km                 NaN    benzin  volkswagen   
freq            32424                 NaN     30107       10687   
mean              NaN            5.723360       NaN         NaN   
std               NaN            3.711984       NaN         NaN   
min               NaN            0.000000       NaN         NaN   
25%               NaN            3.000000       NaN         NaN   
50%               NaN            6.000000       NaN         NaN   
75%               NaN            9.000000       NaN         NaN   
max               NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  no_of_pics   postal_code  \
count              40171                50000     50000.0  50000.000000   
unique                 2                   76         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq               35232                 1946         NaN           NaN   
mean                 NaN                  NaN         0.0  50813.627300   
std                  NaN                  NaN         0.0  25779.747957   
min                  NaN                  NaN         0.0   1067.000000   
25%                  NaN                  NaN         0.0  30451.000000   
50%                  NaN                  NaN         0.0  49577.000000   
75%                  NaN                  NaN         0.0  71540.000000   
max                  NaN                  NaN         0.0  99998.000000   

           last_seen_online  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

1) Columns `seller` and `offer_type` have only one row with a different value except which all the rows have the same value for each of the columns.

2) The column `no_of_pics` has only one value: '0'. 

3) The columns `price`, `kilometers_run` have numeric data stored as text data.

4) The power values for some rows in power_ps column is zero which requires to be changed to 'NA'.

In [148]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)
print(autos["price"].dtype)
autos["kilometers_run"] = autos["kilometers_run"].str.replace(",","")
autos["kilometers_run"] = autos["kilometers_run"].str.replace("km","")
autos["kilometers_run"] = autos["kilometers_run"].astype(int)
print(autos["kilometers_run"].dtype)

int64
int64


In [149]:
autos.rename(columns = {"kilometers_run":"odometer_km"},inplace = True)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  no_of_pics  \
0         peugeot              nein  2016-03-26 00:00:00           0   
1             bmw              nein  2016-04-04 00:00:00           0   
2      volkswagen              nein  2016-03-26 00:00:00           0   
3           smart              nein  2016-03-12 00:00:00           0   
4            ford              nein  2016-04-01 00:00:00           0   
...           ...               ...                  ...         ...   
49995        audi              nein  2016-03-27 00:00:00           0   
49996        opel              nein  2016-03-28 00:00:00           0   
49997        fiat              nein  2016-04-02 00:00:00           0   
49998        audi              nein  2016-03-08 00:00:00           0   
49999        opel              nein  2016-03-13 00:00:00           0   

       postal_code     last_seen_online  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 14:38:50  
...            ...                  ..

In [150]:
autos["odometer_km"].unique().shape
autos["odometer_km"].unique()
autos["odometer_km"].describe()
autos["odometer_km"].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [151]:
autos["price"].unique().shape
autos["price"].describe()
autos["price"].value_counts().head(10)
autos["price"].value_counts().sort_index(ascending = False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64

In [152]:
autos.loc[autos["price"] < 100].sort_values("price").shape

(1762, 20)

In [153]:
autos.loc[autos["price"] >= 10000000].sort_values("price")

date_crawled                                               name  \
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   

       seller offer_type     price   abtest vehicle_type  registration_year  \
11137  privat    Angebot  10000000  control        coupe               1960   
2897   privat    Angebot  11111111     test    limousine               1973   
24384  privat    Angebot  11111111     test          NaN               1995   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
47598  privat    Angebot  12345678  control    limousine               2001   
42221  privat    Angebot  27322222  control    limousine               2014   
39705  privat    Angebot  99999999  control    limousine               1999   

         gearbox  power_PS     model  odometer_km  registration_month  \
11137    manuell       368       NaN       100000                   1   
2897     manuell        48    escort        50000                   3   
24384        NaN         0       NaN       150000                   0   
27371        NaN        95     punto       150000                   0   
39377    manuell        95       v40       150000                   6   
47598    manuell       101    vectra       150000                   3   
42221    manuell       163        c4        40000                   2   
39705  automatik       224  s_klasse       150000                   9   

      fuel_type           brand unrepaired_damage           ad_created  \
11137    benzin  sonstige_autos              nein  2016-03-29 00:00:00   
2897     benzin            ford              nein  2016-03-12 00:00:00   
24384       NaN      volkswagen               NaN  2016-03-21 00:00:00   
27371       NaN            fiat               NaN  2016-03-09 00:00:00   
39377       NaN           volvo              nein  2016-03-08 00:00:00   
47598    benzin            opel              nein  2016-03-31 00:00:00   
42221    diesel         citroen               NaN  2016-03-08 00:00:00   
39705    benzin   mercedes_benz               NaN  2016-03-22 00:00:00   

       no_of_pics  postal_code     last_seen_online  
11137           0        73033  2016-04-06 21:18:11  
2897            0        94469  2016-03-12 22:45:27  
24384           0        18519  2016-03-21 14:40:18  
27371           0        96110  2016-03-09 15:45:47  
39377           0        14542  2016-04-06 23:17:31  
47598           0         4356  2016-03-31 18:56:54  
42221           0        76532  2016-03-08 20:39:05  
39705           0        73525  2016-04-06 05:15:30

In [154]:
#Removing rows with price less than $100.00 and 
#more than $10,000,000 .
autos = autos[autos["price"].between(101, 9999999)].copy()
autos["price"].value_counts().sort_index(ascending = False)

3890000     1
1300000     1
1234566     1
999999      2
999990      1
           ..
120        39
117         1
115         2
111         2
110         3
Name: price, Length: 2314, dtype: int64

From the analysis `price` column:-
1) We see that there are 1762 cars with very low price like less than 100 dollars which is very much unlikely.

2) There are 8 cars with price equal to 10 millon USD. The lowest being a Maserati car with value of 10 million USD. As of Sep 2022 the hight price of a Maserati car in the global market is 2.1 million USD of the model: Maserati MC12 Corsa. So, any price of the used car above 10 million USD is very much unlikely and hence is removed from our dataset. 

In [155]:
print("Distribution of values in the `date_crawled` column in percentage: ")
autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_index() * 100

Distribution of values in the `date_crawled` column in percentage: 


2016-03-05    2.538673
2016-03-06    1.405522
2016-03-07    3.594894
2016-03-08    3.322522
2016-03-09    3.299651
2016-03-10    3.235196
2016-03-11    3.264305
2016-03-12    3.694694
2016-03-13    1.571856
2016-03-14    3.663506
2016-03-15    3.422322
2016-03-16    2.944112
2016-03-17    3.152029
2016-03-18    1.287009
2016-03-19    3.472222
2016-03-20    3.775782
2016-03-21    3.730040
2016-03-22    3.293413
2016-03-23    3.228959
2016-03-24    2.944112
2016-03-25    3.139554
2016-03-26    3.222721
2016-03-27    3.112525
2016-03-28    3.499251
2016-03-29    3.409847
2016-03-30    3.378659
2016-03-31    3.183217
2016-04-01    3.368263
2016-04-02    3.559548
2016-04-03    3.861028
2016-04-04    3.659348
2016-04-05    1.309880
2016-04-06    0.316035
2016-04-07    0.139305
Name: date_crawled, dtype: float64

###### All the information for `date_crawled` was collected from 5th Mar 2016 to  7th Apr 2016. On each of these days about 3 percent of the total information was collected.

In [156]:
print("Distribution of values in the `ad_created` column in percentage: ")
autos["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_index() * 100

Distribution of values in the `ad_created` column in percentage: 


2015-06-11    0.002079
2015-08-10    0.002079
2015-09-09    0.002079
2015-11-10    0.002079
2015-12-05    0.002079
                ...   
2016-04-03    3.885978
2016-04-04    3.694694
2016-04-05    1.183051
2016-04-06    0.324351
2016-04-07    0.124750
Name: ad_created, Length: 76, dtype: float64

###### The `ad_created'` column analysis table has 76 rows from the date 11th Jun 2015 to 7th Apr 2016 i.e in about 10 months or 300 days there were 76 days when new listings were added. 

In [157]:
print("Distribution of values in the `last_seen_online` column in percentage: ")
autos["last_seen_online"].str[:10].value_counts(normalize = True, dropna = False).sort_index() * 100 

Distribution of values in the `last_seen_online` column in percentage: 


2016-03-05     0.108117
2016-03-06     0.430389
2016-03-07     0.538506
2016-03-08     0.729790
2016-03-09     0.956420
2016-03-10     1.064538
2016-03-11     1.241267
2016-03-12     2.382735
2016-03-13     0.887808
2016-03-14     1.262059
2016-03-15     1.582252
2016-03-16     1.642548
2016-03-17     2.806886
2016-03-18     0.731870
2016-03-19     1.576015
2016-03-20     2.066700
2016-03-21     2.056304
2016-03-22     2.137392
2016-03-23     1.858782
2016-03-24     1.971058
2016-03-25     1.910762
2016-03-26     1.669578
2016-03-27     1.551065
2016-03-28     2.081254
2016-03-29     2.233034
2016-03-30     2.463822
2016-03-31     2.384814
2016-04-01     2.287092
2016-04-02     2.486693
2016-04-03     2.513723
2016-04-04     2.451347
2016-04-05    12.518713
2016-04-06    22.195193
2016-04-07    13.221474
Name: last_seen_online, dtype: float64

###### The `last_seen_online` column values were populated from 5th Mar 2016 to 7th Apr 2016 on a regular basis everyday varying in between 0.5 percent to 2.5 percent. However, in the last 3 days the numbers are above 10 perecent whic are unusually high and seems like the data listings were taken down by the management just before closing the business operations. 

In [158]:
autos["registration_year"].describe()
#autos.loc[autos["registration_year"] > 2018, "registration_year"]


count    48096.000000
mean      2004.748752
std         88.012845
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The minimum `registration_year` is 1000 and the maximun `registration_year` is 9999. So, it is clear that there are some data which are not in sync for the `registration_year`.

Because a car can't be first registered after the listing was seen, any vehicle with a `registration_year` above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval.

In [159]:
reg_years = autos[autos["registration_year"].between(1900, 2016)] ["registration_year"].value_counts(normalize = True).sort_index() * 100
reg_yr_70pc = reg_years[reg_years.index > 1999].sum()
print("percentage of car registered since 2000: ", reg_yr_70pc)
reg_yr_90pc = reg_years[reg_years.index > 1995].sum()
print("percentage of car registered since 1996: ", reg_yr_90pc)
reg_yr_99pc = reg_years[reg_years.index > 1978].sum()
print("percentage of car registered since 1980: ", reg_yr_99pc)

percentage of car registered since 2000:  71.9573400250941
percentage of car registered since 1996:  90.23925929130792
percentage of car registered since 1980:  99.0308484402717


From, 1910 to 2016, in these 78 years:
* About, 70 percent of the cars were registered since 2000
* About, 90 percent of the cars were registered since 1996

In [160]:
# List of Top 20 brands with the highest number of cars in the listing of used cars selling.
top_20_brands = (autos["brand"].value_counts(normalize = True) * 100) [:20]
print(top_20_brands)

volkswagen        21.296989
bmw               10.899035
opel              10.809631
mercedes_benz      9.620343
audi               8.614022
ford               6.934049
renault            4.769627
peugeot            2.944112
fiat               2.586494
seat               1.894128
skoda              1.609281
nissan             1.532352
mazda              1.521956
smart              1.434631
citroen            1.413839
toyota             1.270376
hyundai            0.995925
sonstige_autos     0.946025
volvo              0.902362
mini               0.864937
Name: brand, dtype: float64


#### We notice that the top 5 brands in the used car sales listing are all German. Probably, because they make good Cars in Germany and they suit to German people temprament and overall driving conditions. 

#### There are 3 luxury brands in the top 5. That may mean that people who are into luxury cars sell them more often to change into the latest and better models.

<span style="color: red; font-style: italic; font-weight: bold; ">The Top - 20 brands are chosen to be analyzed </span> 

In [161]:
brand_avg_price = {}
for brand in top_20_brands.index:
    price_col = autos[autos["brand"] == brand]["price"]
    avg_price = price_col.mean()
    brand_avg_price[brand] = avg_price

sorted_brand_price = dict(sorted(brand_avg_price.items(), key = lambda item : item[1], reverse = True))
sorted_brand_price

{'sonstige_autos': 23981.927472527474,
 'mini': 10591.985576923076,
 'audi': 9268.353608496258,
 'mercedes_benz': 8582.035011886752,
 'bmw': 8547.199923693246,
 'skoda': 6402.441860465116,
 'volkswagen': 5575.662501220346,
 'hyundai': 5416.23382045929,
 'toyota': 5148.0032733224225,
 'volvo': 4922.767281105991,
 'nissan': 4694.3744911804615,
 'seat': 4357.980241492865,
 'mazda': 4097.042349726776,
 'ford': 4079.83928035982,
 'citroen': 3794.5132352941177,
 'smart': 3538.344927536232,
 'peugeot': 3095.3679378531074,
 'opel': 2988.0351990767454,
 'fiat': 2833.099678456592,
 'renault': 2463.1992153443766}

There are three brands in top-5 list of the avg. selling price  which are also in the top-5 list of highest numbers of cars sold by brand: Audi, Mercedez Benz and BMW.

There are five brands in top-10 list of the avg. selling price which are also in the top-10 list of highest numbers of cars sold by brand: Audi, Mercedez Benz, BMW, Skoda and Volkswagen.

__So, brands with higher average selling pricing are selling more.__

Similarly, the four brands with the lowest average selling price Peugeot, Opel, Fiat and Renault are also in the top-10 list of the highest number of cars being sold by brand; with Opel in third position.

__So, brands with the lowest average selling pricing are also selling more.__

<span style="color :maroon;font-style: italic; font-weight: bold;">It makes sense that the cars within the lowest price range or the highest range are selling more.</span>



In [162]:
top_6_brands = autos["brand"].value_counts().index[:6]
mean_mileage = {}
mean_price = {}

for brand in top_6_brands:
    mileage_col = autos[autos["brand"] == brand]["odometer_km"]
    avg_mileage = mileage_col.mean()
    mean_mileage[brand] = avg_mileage
    
for brand in top_6_brands:
    price_col = autos[autos["brand"] == brand]["price"]
    avg_price = price_col.mean()
    mean_price[brand] = avg_price

mileage_series = pd.Series(mean_mileage)
price_series = pd.Series(mean_price)

mileage_df = pd.DataFrame(mileage_series, columns = ["mileage"])
mileage_df["price"] = price_series
mileage_df.sort_values("price", inplace = True, ascending = False)
print(top_20_brands[:6])
mileage_df


volkswagen       21.296989
bmw              10.899035
opel             10.809631
mercedes_benz     9.620343
audi              8.614022
ford              6.934049
Name: brand, dtype: float64


mileage        price
audi           129619.840695  9268.353608
mercedes_benz  131054.679058  8582.035012
bmw            132849.103396  8547.199924
volkswagen     128995.899639  5575.662501
ford           124334.332834  4079.839280
opel           129438.353530  2988.035199

### Observation:

From our above analysis, we can observe that Audi is the most highly priced car and Opel is the least priced car and both of them have similar mileages. 

Ford has on the average the least mileage among all the brands which is one of the less expensive brands. But, Opel is more than 1000 USD less expensive then Ford but has as good mileage as the most expensive brand Audi.

So, it is not clear that there is any direct correation in betwwn the price of the car and their mileage.

## Data Cleaning Next Steps:

### Step-1 :

__Identify categorical data that uses german words, translate them and map the values to their english counterparts.__

Columns with German words and their English translation :

1. `seller`:- 
        * `privat`: private
    
2. `offer_type`:-
        * `Angebot`: offer 
    
3. `vehicle_type`:-
        * `kleinwagen`: small_cars
        * `kombi`: station_wagon
        * `cabrio`: convertible
        * `andere`: other_types
        
4. `gearbox`:-
        * `manuell`: manual
        * `automatik`: automatic
        
5. `fuel_type`:-
        * `benzin`: petrol
        * `electro`: electric
        
6.  `unrepaired_damage`:-
        * `nein`: no
        * `ja`: yes

By, checking manually with each column and using the unique() method to find the unique values for those coulumns we found there are 6 columns with German words. Next, we are going to map them to their English counterparts. 

In [163]:
# Creating the mapping dictionary from 'German' to 'english' for all the columns reqired 
map_seller = {"privat":"private"}
map_offer_type = {"Angebot":"offer"}
map_vehicle_type = {
                        "kleinwagen": "small_cars",
                        "kombi": "station_wagon",
                        "cabrio": "convertible",
                        "andere": "other_types"
                    }
map_gearbox = {
                   "manuell": "manual",
                   "automatik": "automatic"
              }
map_fuel_type = {
                    "benzin": "petrol",
                    "electro": "electric"
                }

map_unrepaired_damage = {
                    "nein": "no",
                    "ja": "yes"
                }

#autos.loc[:,"seller"].replace(map_seller, inplace = True)
#autos.loc[:,"offer_type"].replace(map_offer_type, inplace = True)
#autos.loc[:,"vehicle_type"].replace(map_vehicle_type, inplace = True)
#autos.loc[:,"gearbox"].replace(map_gearbox, inplace = True)
#autos.loc[:,"fuel_type"].replace(map_fuel_type, inplace = True)
#autos.loc[:,"unrepaired_damage"].replace(map_unrepaired_damage, inplace = True)

autos.loc[:,"seller"] = autos.loc[:,"seller"].map(map_seller).copy()
autos.loc[:,"offer_type"] = autos.loc[:,"offer_type"].map(map_offer_type).copy()
autos.loc[:,"vehicle_type"] = autos.loc[:,"vehicle_type"].map(map_vehicle_type).copy()
autos.loc[:,"gearbox"] = autos.loc[:,"gearbox"].map(map_gearbox).copy()
autos.loc[:,"fuel_type"] = autos.loc[:,"fuel_type"].map(map_fuel_type).copy()
autos.loc[:,"unrepaired_damage"] = autos.loc[:,"unrepaired_damage"].map(map_unrepaired_damage).copy()
autos


date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

        seller offer_type  price   abtest   vehicle_type  registration_year  \
0      private      offer   5000  control            NaN               2004   
1      private      offer   8500  control            NaN               1997   
2      private      offer   8990     test            NaN               2009   
3      private      offer   4350  control     small_cars               2007   
4      private      offer   1350     test  station_wagon               2003   
...        ...        ...    ...      ...            ...                ...   
49995  private      offer  24900  control            NaN               2011   
49996  private      offer   1980  control    convertible               1996   
49997  private      offer  13200     test    convertible               2014   
49998  private      offer  22900  control  station_wagon               2013   
49999  private      offer   1250  control            NaN               1996   

         gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0         manual       158  andere       150000                   3       NaN   
1      automatic       286     7er       150000                   6    petrol   
2         manual       102    golf        70000                   7    petrol   
3      automatic        71  fortwo        70000                   6    petrol   
4         manual         0   focus       150000                   7    petrol   
...          ...       ...     ...          ...                 ...       ...   
49995  automatic       239      q5       100000                   1       NaN   
49996     manual        75   astra       150000                   5    petrol   
49997  automatic        69     500         5000                  11    petrol   
49998     manual       150      a3        40000                  11       NaN   
49999     manual       101  vectra       150000                   1    petrol   

            brand unrepaired_damage           ad_created  no_of_pics  \
0         peugeot                no  2016-03-26 00:00:00           0   
1             bmw                no  2016-04-04 00:00:00           0   
2      volkswagen                no  2016-03-26 00:00:00           0   
3           smart                no  2016-03-12 00:00:00           0   
4            ford                no  2016-04-01 00:00:00           0   
...           ...               ...                  ...         ...   
49995        audi                no  2016-03-27 00:00:00           0   
49996        opel                no  2016-03-28 00:00:00           0   
49997        fiat                no  2016-04-02 00:00:00           0   
49998        audi                no  2016-03-08 00:00:00           0   
49999        opel                no  2016-03-13 00:00:00           0   

       postal_code     last_seen_online  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 14:38:50  
..

### Step-2 :
__Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321__

There are three columns with "Date" values:
1. date_crawled
2. ad_created
3. last_seen_online

> _we can change the date string in the format "2016-03-21 17:47:46" to "20160321" by replacing "-" with empty strings "" and removing the time string after that._


In [164]:
#date_str = autos.loc[:,"date_crawled"].str[:10]
#date_str = date_str.str.replace("-","")
#print(autos.loc[:,"date_crawled"].dtype)
#print(autos.loc[:,"date_crawled"])



autos.loc[:,"date_crawled"] = autos.loc[:,"date_crawled"].str[:10].str.replace("-","").astype(int)
autos.loc[:,"ad_created"] = autos.loc[:,"ad_created"].str[:10].str.replace("-","").astype(int)
autos.loc[:,"last_seen_online"] = autos.loc[:,"last_seen_online"].str[:10].str.replace("-","").astype(int)
print(autos.loc[:,"last_seen_online"].dtype)
autos

int64


date_crawled                                               name  \
0          20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   
1          20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2          20160326                         Volkswagen_Golf_1.6_United   
3          20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4          20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...             ...                                                ...   
49995      20160327   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996      20160328  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997      20160402                     Fiat_500_C_1.2_Dualogic_Lounge   
49998      20160308                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999      20160314                                Opel_Vectra_1.6_16V   

        seller offer_type  price   abtest   vehicle_type  registration_year  \
0      private      offer   5000  control            NaN               2004   
1      private      offer   8500  control            NaN               1997   
2      private      offer   8990     test            NaN               2009   
3      private      offer   4350  control     small_cars               2007   
4      private      offer   1350     test  station_wagon               2003   
...        ...        ...    ...      ...            ...                ...   
49995  private      offer  24900  control            NaN               2011   
49996  private      offer   1980  control    convertible               1996   
49997  private      offer  13200     test    convertible               2014   
49998  private      offer  22900  control  station_wagon               2013   
49999  private      offer   1250  control            NaN               1996   

         gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0         manual       158  andere       150000                   3       NaN   
1      automatic       286     7er       150000                   6    petrol   
2         manual       102    golf        70000                   7    petrol   
3      automatic        71  fortwo        70000                   6    petrol   
4         manual         0   focus       150000                   7    petrol   
...          ...       ...     ...          ...                 ...       ...   
49995  automatic       239      q5       100000                   1       NaN   
49996     manual        75   astra       150000                   5    petrol   
49997  automatic        69     500         5000                  11    petrol   
49998     manual       150      a3        40000                  11       NaN   
49999     manual       101  vectra       150000                   1    petrol   

            brand unrepaired_damage  ad_created  no_of_pics  postal_code  \
0         peugeot                no    20160326           0        79588   
1             bmw                no    20160404           0        71034   
2      volkswagen                no    20160326           0        35394   
3           smart                no    20160312           0        33729   
4            ford                no    20160401           0        39218   
...           ...               ...         ...         ...          ...   
49995        audi                no    20160327           0        82131   
49996        opel                no    20160328           0        44807   
49997        fiat                no    20160402           0        73430   
49998        audi                no    20160308           0        35683   
49999        opel                no    20160313           0        45897   

       last_seen_online  
0              20160406  
1              20160406  
2              20160406  
3              20160315  
4              20160401  
...                 ...  
49995          20160401  
49996          20160402  
49997          20160404  
49998          20160405

### Step-3 :
__See if there are particular keywords in the name column that you can extract as new columns:__



In [165]:
# Checking for cars with "sport" in their name.
sport_utility = autos[autos["name"].str.contains("Sport", case= False)]
sport_utility



date_crawled                                               name  \
24         20160403                         BMW_535i_xDrive_Sport_Aut.   
33         20160315  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
48         20160320                           Mercedes_C180_Sport_1_8l   
50         20160309  •??_????????_??•Top!!!_Peugeot_308_sw_Sport_•?...   
83         20160405                        BMW_118d_M_Sport_Packet_DPF   
...             ...                                                ...   
49955      20160330                     Golf_1.9_TDI_DPF_4mot_GT_Sport   
49961      20160314            Golf_2.0_TDI_Sportline_Sonderproduktion   
49969      20160317             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49992      20160310              Fiat_Grande_Punto_1.4_T_Jet_16V_Sport   
49998      20160308                 Audi_A3_2.0_TDI_Sportback_Ambition   

        seller offer_type  price   abtest   vehicle_type  registration_year  \
24     private      offer  48500  control            NaN               2014   
33     private      offer   4800     test  station_wagon               2006   
48     private      offer   1099     test            NaN               1998   
50     private      offer   5999     test  station_wagon               2008   
83     private      offer   8490     test            NaN               2008   
...        ...        ...    ...      ...            ...                ...   
49955  private      offer   3400  control            NaN               2008   
49961  private      offer   3200  control            NaN               2007   
49969  private      offer   4500  control            NaN               2005   
49992  private      offer   4800  control     small_cars               2009   
49998  private      offer  22900  control  station_wagon               2013   

         gearbox  power_PS     model  odometer_km  registration_month  \
24     automatic       306       5er        30000                  12   
33        manual       140        a3       150000                   8   
48        manual       122  c_klasse       150000                   2   
50        manual       150   3_reihe       125000                   8   
83        manual       143       1er       150000                  12   
...          ...       ...       ...          ...                 ...   
49955        NaN         0      golf       150000                  11   
49961     manual       140      golf       150000                   7   
49969     manual       136   x_trail       150000                   5   
49992     manual       120    andere       125000                   9   
49998     manual       150        a3        40000                  11   

      fuel_type          brand unrepaired_damage  ad_created  no_of_pics  \
24       petrol            bmw                no    20160403           0   
33          NaN           audi                no    20160315           0   
48       petrol  mercedes_benz                no    20160320           0   
50       petrol        peugeot                no    20160309           0   
83          NaN            bmw                no    20160405           0   
...         ...            ...               ...         ...         ...   
49955       NaN     volkswagen                no    20160330           0   
49961       NaN     volkswagen               NaN    20160314           0   
49969       NaN         nissan                no    20160317           0   
49992       NaN           fiat                no    20160310           0   
49998       NaN           audi                no    20160308           0   

       postal_code  last_seen_online  
24           22547          20160407  
33           61169          20160407  
48           83024          20160324  
50           55569          20160407  
83           81825          20160405  
...            ...               ...  
49955        83512          20160330  
49961        65760          20160314  
49969        17379          

__There are 2593 rows in the name column with the keyword "sport". We can create a separate column named "sport_utility" and assign 'yes' or 'no'__

In [167]:
autos.loc[:,"sport_utility"] = "no"
autos.loc[autos["name"].str.contains("sport", case= False),"sport_utility"] = "Yes"
autos

date_crawled                                               name  \
0          20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   
1          20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2          20160326                         Volkswagen_Golf_1.6_United   
3          20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4          20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...             ...                                                ...   
49995      20160327   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996      20160328  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997      20160402                     Fiat_500_C_1.2_Dualogic_Lounge   
49998      20160308                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999      20160314                                Opel_Vectra_1.6_16V   

        seller offer_type  price   abtest   vehicle_type  registration_year  \
0      private      offer   5000  control            NaN               2004   
1      private      offer   8500  control            NaN               1997   
2      private      offer   8990     test            NaN               2009   
3      private      offer   4350  control     small_cars               2007   
4      private      offer   1350     test  station_wagon               2003   
...        ...        ...    ...      ...            ...                ...   
49995  private      offer  24900  control            NaN               2011   
49996  private      offer   1980  control    convertible               1996   
49997  private      offer  13200     test    convertible               2014   
49998  private      offer  22900  control  station_wagon               2013   
49999  private      offer   1250  control            NaN               1996   

         gearbox  power_PS  ... odometer_km  registration_month  fuel_type  \
0         manual       158  ...      150000                   3        NaN   
1      automatic       286  ...      150000                   6     petrol   
2         manual       102  ...       70000                   7     petrol   
3      automatic        71  ...       70000                   6     petrol   
4         manual         0  ...      150000                   7     petrol   
...          ...       ...  ...         ...                 ...        ...   
49995  automatic       239  ...      100000                   1        NaN   
49996     manual        75  ...      150000                   5     petrol   
49997  automatic        69  ...        5000                  11     petrol   
49998     manual       150  ...       40000                  11        NaN   
49999     manual       101  ...      150000                   1     petrol   

            brand unrepaired_damage ad_created  no_of_pics  postal_code  \
0         peugeot                no   20160326           0        79588   
1             bmw                no   20160404           0        71034   
2      volkswagen                no   20160326           0        35394   
3           smart                no   20160312           0        33729   
4            ford                no   20160401           0        39218   
...           ...               ...        ...         ...          ...   
49995        audi                no   20160327           0        82131   
49996        opel                no   20160328           0        44807   
49997        fiat                no   20160402           0        73430   
49998        audi                no   20160308           0        35683   
49999        opel                no   20160313           0        45897   

       last_seen_online  sport_utility  
0              20160406             no  
1              20160406             no  
2              20160406             no  
3              20160315             no  
4              20160401             no  
...                 ...            ...  
49995          20160401             no  
499

__Looking for vehicle names with "wohnmobil" or "Wohnwagen" whic are German for "Motor Home" and "Caravan":__

In [177]:

autos[autos["name"].str.contains("wohnmobil", case= False) | autos["name"].str.contains("Wohnwagen", case= False) ]



date_crawled                                               name  \
362        20160402  Mercedes_Benz_Viano_CDI_3.0_Marco_Polo_Wohnmob...   
1932       20160306  Verkaufe_unser_Wohnmobilbegleitfahrzeug_mit_Tr...   
2403       20160324                  Fiat_Ducato__wohnmobil_.2.5diesel   
3940       20160315                            Volkswagen_T5_Wohnmobil   
4443       20160316      Mercedes_MB_100_Diesel_Wohnmobil_Luna_Holiday   
8555       20160402  Barkas_B1000__2Takter_Fensterbus__Partybus__Wo...   
10335      20160324                            Volkswagen_T4_Wohnmobil   
12066      20160310                             Wohnwagen_LMC__10_2009   
13894      20160324                VW_T3_Bulli_Bj_80_Camping_Wohnmobil   
14189      20160402  300_TDI_Land_Rover_Defender_Wohnmobil_Allrad_4...   
14373      20160329  Wohnmobil__Fahrer_aufgepasst!__Smart_auf_Anhae...   
19370      20160402  VW_T4_Multivan_Bus___TÜV___VR6_mit_140_PS___LP...   
21869      20160404                                    VW_T3_Wohnmobil   
23449      20160326              Sprinter_4x4_Werkstattwagen_Wohnmobil   
23785      20160330  Wohnmobil_fuer4_Personen__wegen_Platzmangel_un...   
25867      20160320  vw_t4_multivan_allstar_2.5_benzin_lpg_wohnwage...   
27600      20160329                    BMW_e39_Tausch_gegen_Wohnmobile   
31238      20160316  vw_t4_multivan_allstar_2.5_benzin_lpg_wohnwage...   
32101      20160329  Volkswagen_VW_T4_Wohnmobil___VR6___140_PS___Au...   
33098      20160329              T4_Multivan_151_PS_Wohnmobilzulassung   
34491      20160324  VW_T4_Multivan_Wohnmobilzulassung_Reimo_Ausste...   
35143      20160314    VW_T5_Multivan_Facelift_mit_Wohnmobil_Zulassung   
40600      20160309  VW_T4_Transporter_Wohnmobil_•_VR6_•_140_PS_•_g...   
41929      20160401  Volkswagen_VW_T4_California_mit_Westfalia_Auss...   
44144      20160316      Mercedes_Benz_MB_100_Wohnmobil_Laluna_Holiday   
45118      20160324                   Peugeot_Wohnmobil_Europa_650_FFB   
48846      20160311  Passat_35i_+_Oldtimer_Wohnwagen_zum_tausch_geg...   

        seller offer_type  price   abtest   vehicle_type  registration_year  \
362    private      offer  42999  control            NaN               2013   
1932   private      offer   2900     test    convertible               2002   
2403   private      offer   3999     test            NaN               1987   
3940   private      offer  19500  control            NaN               2010   
4443   private      offer   8900  control            NaN               1992   
8555   private      offer   3900  control            NaN               1988   
10335  private      offer   2899     test            NaN               1994   
12066  private      offer   9700  control    other_types               2009   
13894  private      offer   1800     test            NaN               1980   
14189  private      offer  16800     test            NaN               1996   
14373  private      offer   5999  control     small_cars               2005   
19370  private      offer   4700  control            NaN               1996   
21869  private      offer   6999     test    other_types               1991   
23449  private      offer  17500  control    other_types               2007   
23785  private      offer   1750  control            NaN               1995   
25867  private      offer   4950     test            NaN               1996   
27600  private      offer   6000  control  station_wagon               2003   
31238  private      offer   5700  control            NaN               1996   
32101  private      offer  10500  control            NaN               1998   
33098  private      offer   6000  control            NaN               1999   
34491  private      offer   1500     test            NaN               1992   
35143  private      offer  31900     test            NaN               2010   
40600  private      offer  10500  control            NaN               1998   
41929  private      offer   9600  cont

There are 27 rows with the term "wohnmobil" or "Wohnwagen" in the name column. So we can create a new column named "motor_home" with values "yes" for the corresponding rows.

In [182]:
autos["motor_home"] = "no"
autos.loc[(autos["name"].str.contains("wohnmobil", case= False) | autos["name"].str.contains("Wohnwagen", case= False)), "motor_home"] = "yes"
autos[autos["name"].str.contains("wohnmobil", case= False) | autos["name"].str.contains("Wohnwagen", case= False) ]

date_crawled                                               name  \
362        20160402  Mercedes_Benz_Viano_CDI_3.0_Marco_Polo_Wohnmob...   
1932       20160306  Verkaufe_unser_Wohnmobilbegleitfahrzeug_mit_Tr...   
2403       20160324                  Fiat_Ducato__wohnmobil_.2.5diesel   
3940       20160315                            Volkswagen_T5_Wohnmobil   
4443       20160316      Mercedes_MB_100_Diesel_Wohnmobil_Luna_Holiday   
8555       20160402  Barkas_B1000__2Takter_Fensterbus__Partybus__Wo...   
10335      20160324                            Volkswagen_T4_Wohnmobil   
12066      20160310                             Wohnwagen_LMC__10_2009   
13894      20160324                VW_T3_Bulli_Bj_80_Camping_Wohnmobil   
14189      20160402  300_TDI_Land_Rover_Defender_Wohnmobil_Allrad_4...   
14373      20160329  Wohnmobil__Fahrer_aufgepasst!__Smart_auf_Anhae...   
19370      20160402  VW_T4_Multivan_Bus___TÜV___VR6_mit_140_PS___LP...   
21869      20160404                                    VW_T3_Wohnmobil   
23449      20160326              Sprinter_4x4_Werkstattwagen_Wohnmobil   
23785      20160330  Wohnmobil_fuer4_Personen__wegen_Platzmangel_un...   
25867      20160320  vw_t4_multivan_allstar_2.5_benzin_lpg_wohnwage...   
27600      20160329                    BMW_e39_Tausch_gegen_Wohnmobile   
31238      20160316  vw_t4_multivan_allstar_2.5_benzin_lpg_wohnwage...   
32101      20160329  Volkswagen_VW_T4_Wohnmobil___VR6___140_PS___Au...   
33098      20160329              T4_Multivan_151_PS_Wohnmobilzulassung   
34491      20160324  VW_T4_Multivan_Wohnmobilzulassung_Reimo_Ausste...   
35143      20160314    VW_T5_Multivan_Facelift_mit_Wohnmobil_Zulassung   
40600      20160309  VW_T4_Transporter_Wohnmobil_•_VR6_•_140_PS_•_g...   
41929      20160401  Volkswagen_VW_T4_California_mit_Westfalia_Auss...   
44144      20160316      Mercedes_Benz_MB_100_Wohnmobil_Laluna_Holiday   
45118      20160324                   Peugeot_Wohnmobil_Europa_650_FFB   
48846      20160311  Passat_35i_+_Oldtimer_Wohnwagen_zum_tausch_geg...   

        seller offer_type  price   abtest   vehicle_type  registration_year  \
362    private      offer  42999  control            NaN               2013   
1932   private      offer   2900     test    convertible               2002   
2403   private      offer   3999     test            NaN               1987   
3940   private      offer  19500  control            NaN               2010   
4443   private      offer   8900  control            NaN               1992   
8555   private      offer   3900  control            NaN               1988   
10335  private      offer   2899     test            NaN               1994   
12066  private      offer   9700  control    other_types               2009   
13894  private      offer   1800     test            NaN               1980   
14189  private      offer  16800     test            NaN               1996   
14373  private      offer   5999  control     small_cars               2005   
19370  private      offer   4700  control            NaN               1996   
21869  private      offer   6999     test    other_types               1991   
23449  private      offer  17500  control    other_types               2007   
23785  private      offer   1750  control            NaN               1995   
25867  private      offer   4950     test            NaN               1996   
27600  private      offer   6000  control  station_wagon               2003   
31238  private      offer   5700  control            NaN               1996   
32101  private      offer  10500  control            NaN               1998   
33098  private      offer   6000  control            NaN               1999   
34491  private      offer   1500     test            NaN               1992   
35143  private      offer  31900     test            NaN               2010   
40600  private      offer  10500  control            NaN               1998   
41929  private      offer   9600  cont

## Analysis Next Steps:

### Step-1 :

__Find the most common brand model combination__

In [196]:
# Joining the `brand` and `model column together with '_' 
# in between for better readanility.

brand_model = autos["brand"] + "_" + autos["model"]
brand_model.describe()


count               45718
unique                291
top       volkswagen_golf
freq                 3864
dtype: object

_From, the above analysis it is clear that 'Volkswagen' brand and 'Polo' model are the most common brand model combination with 3,864 occurances in the Listings._ 

## Analysis Next Steps:

### Step-2 :

__Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.__

In [201]:
autos["odometer_km"].unique()


array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

There are 13 unique values for mileage with the lowest being 5,000  and the highest is 150,000.

We can divide them into 3 groups:

1. Mileage <= 50,000 

2. 50,000 < Mileage <= 100,000

3. Mileage > 100,000



In [219]:
low_mile_avg_price = autos.loc[autos["odometer_km"] <= 50000, "price" ].mean()
mid_mile_avg_price = autos.loc[autos["odometer_km"].between(50001,100000), "price"].mean()
high_mile_avg_price = autos.loc[autos["odometer_km"] > 100000, "price"].mean()

print("Mileage <= 50,000 : ", low_mile_avg_price, "\n")
print("50,000 < Mileage <= 100,000 : ", mid_mile_avg_price, "\n")
print("Mileage > 100,000 : ", high_mile_avg_price, "\n")

Mileage <= 50,000 :  16234.011959521618 

50,000 < Mileage <= 100,000 :  9530.125641362978 

Mileage > 100,000 :  4215.382106398872 



__It is clear from above analysis that the group with the lowest price has highest mileage, the middle price range group has less mileage and the highest priced group has the the least mileage.__

## Analysis Next Steps:

### Step-3 :

__How much cheaper are cars with damage than their non-damaged counterparts?.__

In [228]:
# Avg. Price of damaged cars
damaged_cars_prc = autos.loc[autos["unrepaired_damage"] == "yes","price"].mean()
# Avg. Price of non-damaged cars
undamaged_cars_prc = autos.loc[autos["unrepaired_damage"] == "no","price"].mean()

print("Difference between the price of non-damaged and damaged cars :")
print(undamaged_cars_prc - damaged_cars_prc)

Difference between the price of non-damaged and damaged cars :
4791.142710620039


__So, the damaged cars are about 5000 USD cheaper then their non-damaged counterparts.__